style.md

```diff
@@ -1557,7 +1557,237 @@ type Client struct {
+b.Write(...) // ok
+```
+
+</td></tr>
+</tbody></table>
+
+#### Embedding Should Not Change Type Semantics
+
+An embedded (inner) type should not add semantically unrelated or extraneous
+methods or fields to the embedding (outer) type's exported API, nor should it
+add methods or fields that imply that the type behaves other than explicitly
+defined. This applies both to newly-available (via embedding) and "shadowed"
+methods and fields.
+
+<table>
+<thead><tr><th>Bad</th><th>Good</th></tr></thead>
+<tbody>
+<tr><td>
+
+```go
+type Car struct {
+    io.ReadWriteCloser
+    sync.Locker
+
+    // other fields
+}
+
+// Close closes the indicated door.
+func (c *Car) Close(d Door) error {
+    c.Locker.Lock()
+    defer c.Locker.Unlock()
+
+    // ...
+}
+
+// Lock locks c's doors.
+func (c *Car) Lock() error {
+    c.Locker.Lock()
+    defer c.Locker.Unlock()
+
+    // ...
+}
+
+// Unlock unlocks c's doors.
+func (c *Car) Unlock() error {
+    c.Locker.Lock()
+    defer c.Locker.Unlock()
+
+    // ...
+}
+
+// later, in user code...
+
+c.Lock()        // Bad: unclear intent
+c.Locker.Lock() // Bad: accesses internals
+c.Read(...)     // Bad: Car is not semantically an io.Reader
+```
+
+</td><td>
+
+```go
+type Car struct {
+    buf    bytes.Buffer
+    locker sync.Locker
+    // other fields
+}
+
+// Close closes the indicated door.
+func (c *Car) Close(d Door) error {
+    // ...
+}
+
+// Lock locks c's doors.
+func (c *Car) Lock() error {
+    c.locker.Lock()
+    defer c.locker.Unlock()
+
+    // ...
+}
+
+// later, in user code...
+
+c.Lock()        // Good: unambiguous
+c.locker.Lock() // Good: is not visible to users
+c.buf.Read()    // Good: is not visible to users
+```
+
+</td></tr>
+</tbody></table>
+
+#### Embedding Should Not Expose Implementation Details
+
+An embedded type should not expose controls, mechanics, internal details, or
+other information/access outside of what is already explicitly available
+through the embedding type's API. In those cases, the inner type should be an
+unexported field instead.
+
+<table>
+<thead><tr><th>Bad</th><th>Good</th></tr></thead>
+<tbody>
+<tr><td>
+
+```go
+type Client struct {
+    sync.Mutex
+    sync.WaitGroup
+    bytes.Buffer
+    url.URL
+}
+```
+
+</td><td>
+
+```go
+type Client struct {
+    mtx sync.Mutex
+    wg  sync.WaitGroup
+    buf bytes.Buffer
+    url url.URL
+}
+```
+
+</td></tr>
+</tbody></table>
+
+#### Embedding Values vs Pointers vs Interfaces
+
+Prefer embedding value types where possible (if the type's zero value is
+useful); otherwise, always embed a canonically-constructed type (the type
+returned by its constructor(s)). Never share embedded pointers between outer
+values.
+
+<table>
+<thead><tr><th>Bad</th><th>Good</th></tr></thead>
+<tbody>
+<tr><td>
+
+```go
+type Foo struct {
+    *bytes.Buffer // Bad: has useful zero value
+    time.Ticker   // Bad: non-canonical
+}
+```
+
+</td><td>
+
+```go
+type Foo struct {
+    bytes.Buffer
+    *time.Ticker
```

Comment:
hmm I'm not sure about this recommendation, since `time.Ticker` is probably fine if we always use `*Foo` everywhere?

Reply:
It's fine if `*Foo` is the canonically-constructed version of "a Foo". If the constructor returns `Foo` instead of `*Foo`, that's probably less okay.

I need to rework these examples - agreed that they're not great.